cmd/inspect/inspect_policy_data.go

```diff
@@ -0,0 +1,159 @@
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+// Define the `ec inspect policy-data` command
+package inspect
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/fs"
+	"path/filepath"
+	"strings"
+
+	hd "github.com/MakeNowJust/heredoc"
+	"github.com/ghodss/yaml"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/afero"
+	"github.com/spf13/cobra"
+	"golang.org/x/exp/slices"
+
+	"github.com/hacbs-contract/ec-cli/internal/policy/source"
+	"github.com/hacbs-contract/ec-cli/internal/utils"
+)
+
+func inspectPolicyDataCmd() *cobra.Command {
+	var (
+		sourceUrls   []string
+		destDir      string
+		outputFormat string
+	)
+
+	validFormats := []string{"json", "yaml"}
+
+	cmd := &cobra.Command{
+		Use:   "policy-data --source <source-url>",
+		Short: "Read policy data from source urls and displays the data",
+
+		Long: hd.Doc(`
+			Read policy data from source urls and displays the data.
+
+			This fetches policy sources similar to the 'ec fetch policy' command, but once
+			the policy is fetched it reads json and yaml files inside the policy source and
+			displays the data.
+
+			Note that this command is not typically required to verify the Enterprise
+			Contract. It has been made available for troubleshooting and debugging purposes.
+		`),
+
+		Example: hd.Doc(`
+			Print data from a given source url:
+
+			  ec inspect policy-data --source quay.io/hacbs-contract/ec-policy-data
+		`),
+
+		Args: cobra.NoArgs,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if !slices.Contains(validFormats, outputFormat) {
+				return fmt.Errorf("invalid value for --output '%s'. accepted values: %s", outputFormat, strings.Join(validFormats, ", "))
+			}
+
+			if destDir == "" {
+				workDir, err := utils.CreateWorkDir(afero.NewOsFs())
+				if err != nil {
+					log.Debug("Failed to create work dir!")
+					return err
+				}
+				destDir = workDir
+			}
+
+			allData := map[string]interface{}{}
+
+			for _, url := range sourceUrls {
+				s := &source.PolicyUrl{Url: url, Kind: source.PolicyKind}
+
+				// Download
+				policyDir, err := s.GetPolicy(cmd.Context(), destDir, false)
+				if err != nil {
+					return err
+				}
+
+				err = afero.Walk(afero.NewOsFs(), policyDir, func(path string, d fs.FileInfo, readErr error) error {
+					if readErr != nil {
+						return readErr
+					}
+
+					if d.IsDir() {
+						return nil
+					}
+
+					fileExt := strings.ToLower(filepath.Ext(path))
+					if slices.Contains([]string{".yaml", ".yml", ".json"}, fileExt) {
+						log.Debugf("Found data file %s", path)
+
+						contents, err := afero.ReadFile(afero.NewOsFs(), path)
+						if err != nil {
+							return nil
+						}
+
+						yamlData := make(map[string]interface{})
+
+						// Should work for both yaml and json
+						err = yaml.Unmarshal(contents, &yamlData)
+						if err != nil {
+							return err
+						}
+
+						// If we find a duplicate key, throw an error. I think that is what Conftest would do.
```

Comment:
There is a [bundle.Merge](https://pkg.go.dev/github.com/open-policy-agent/opa@v0.50.0/bundle#Merge) function in OPA, I wonder if we should use that. The downside is that we would have two ways of loading data. I do assume that Conftest delegates to this in the end. Something like this might work (from the top of my head):

```go
// in the loop over bundles
var complete *bundle.Bundle
r := bundle.NewCustomReader(bundle.NewDirectoryLoader(dir))
b, err := r.Read()

if complete == nil {
	complete = b
} else {
	complete.Merge(b)
}
// dump in YAML
yaml.Marshal(complete.Data)
```

Reply:
I think actually Conftest does its own style of data loading different from OPA. In particular, for OPA the directory structure is significant, so you get nested keys for files in nested directories, but Conftest just flattens everything IIUC. I was poking around [here](https://github.com/open-policy-agent/conftest/blob/master/builtins/parse_config.go) trying to understand this.

That said I might experiment with this and see what we get.

Reply:
We don't want to diverge from Conftest, that would be confusing and work in vain...